### PR TITLE
Device Primary IP now accepts IP Addresses on Modules.

### LINKED ISSUE #5791: Allow device primary IP from interfaces in child modules
Follow-up to #2101. Need to allow devices to set primary IP address from interfaces that are members of their child modules

### PR DESCRIPTION
**Pull request by smk4664**
# Closes #5791 #5792 
# What's Changed

Device Primary IP now accepts IP Addresses from Modules.
Device Components from Modules show up in the Device Component Tab (and count).

# Screenshots

OLD: 
image
image

NEW:
image
image

# TODO

- [ ] Explanation of Change(s)
- [ ] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- [ ] Attached Screenshots, Payload Example
- [ ] Unit, Integration Tests
- [ ] Documentation Updates (when adding/changing features)
- [ ] Example App Updates (when adding/changing features)
- [ ] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: changes/5792.added [hunk 1]
```diff
@@ -0,0 +1 @@
+Device Component tabs now show components from attached modules.
```
Comment (glennmatthews) [inline changes/5792.added:1]: Perhaps:

```suggestion
Added display of components from installed modules to the Device Component tabs.
```

File: changes/5791.changed [hunk 2]
```diff
@@ -0,0 +1 @@
+Allow Device Primary IP can now be selected from Child Modules.
```
Comment (glennmatthews) [inline changes/5791.changed:1]: Perhaps `.added` rather than `.changed` since it's net-new functionality. Perhaps:

```suggestion
Added support for specifying a Device's Primary IP from an interface on a child Module.
```

File: nautobot/dcim/forms.py [hunk 3]
```diff
@@ -3774,6 +3774,9 @@ class InventoryItemFilterForm(DeviceComponentFilterForm):
 #
 
 
+# TODO: Cable connection forms need to support components within Modules
+# While it allows connecting components from within Modules, you cannot
+# filter by modules.
```
Comment (glennmatthews) [inline nautobot/dcim/forms.py:3779]: This is #5790 right?

File: nautobot/dcim/models/devices.py [hunk 4]
```diff
@@ -744,7 +744,7 @@ def clean(self):
                 pass
 
         # Validate primary IP addresses
-        vc_interfaces = self.vc_interfaces.all()
+        interfaces = self.all_interfaces.all()
```
Comment (glennmatthews) [inline nautobot/dcim/models/devices.py:747]: maybe `all_interfaces` just to be consistent?

File: nautobot/dcim/models/devices.py [hunk 5]
```diff
@@ -916,6 +916,7 @@ def vc_interfaces(self):
             filter_q |= Q(device__virtual_chassis=self.virtual_chassis, mgmt_only=False)
         return Interface.objects.filter(filter_q)
 
+    # TODO - I am not sure if this should change with the introduction of modules
```
Comment (glennmatthews) [inline nautobot/dcim/models/devices.py:919]: Yeah probably `VirtualChassis.member_interfaces` needs to roll up from `Device.all_interfaces()` as well?

File: nautobot/dcim/models/devices.py [hunk 6]
```diff
@@ -955,6 +956,55 @@ def get_children(self):
         """
         return Device.objects.filter(parent_bay__device=self.pk)
 
+    @property
+    def modules(self):
```
Comment (glennmatthews) [inline nautobot/dcim/models/devices.py:960]: Maybe `all_modules` to make it clearer that this includes descendants recursively?

File: nautobot/dcim/models/devices.py [hunk 7]
```diff
@@ -955,6 +956,55 @@ def get_children(self):
...
+    def all_front_ports(self):
+        return FrontPort.objects.filter(models.Q(device=self) | models.Q(module__in=self.modules))
+
+    @property
+    def all_interfaces(self):
+        return self.vc_interfaces | Interface.objects.filter(module__in=self.modules)
+
+    @property
+    def all_module_bays(self):
+        return ModuleBay.objects.filter(models.Q(parent_device=self) | models.Q(parent_device__in=self.modules))
```
Comment (glennmatthews) [inline nautobot/dcim/models/devices.py:994]: Is this one right? `parent_device__in=self.modules` looks sketchy to me.

File: nautobot/dcim/models/devices.py [hunk 8]
```diff
@@ -955,6 +956,55 @@ def get_children(self):
...
+    def all_power_ports(self):
+        return PowerPort.objects.filter(models.Q(device=self) | models.Q(module__in=self.modules))
+
+    @property
+    def all_power_outlets(self):
+        return PowerOutlet.objects.filter(models.Q(device=self) | models.Q(module__in=self.modules))
+
+    @property
+    def all_rear_ports(self):
+        return RearPort.objects.filter(device=self) | RearPort.objects.filter(module__in=self.modules)
```
Comment (glennmatthews) [inline nautobot/dcim/models/devices.py:1006]: Any behavioral or performance difference between these two patterns? `filter(x) | filter(y)` vs `filter(Q(x) | Q(y))`? We should pick one and be self-consistent.

File: nautobot/dcim/models/devices.py [hunk 9]
```diff
@@ -1752,4 +1803,4 @@ def get_children(self):
         """
         Return the set of child Modules installed in ModuleBays within this Module.
         """
-        return Module.objects.filter(parent_module_bay__parent_module=self)
+        return Module.objects.filter(parent_module_bay__parent_module=self.pk)
```
Comment (glennmatthews) [inline nautobot/dcim/models/devices.py:1806]: Why this change?

File: nautobot/dcim/models/devices.py [hunk 10]
```diff
@@ -955,6 +956,55 @@ def get_children(self):
...
+        recursion_depth = 4
+        qs = Module.objects.all()
+        query = models.Q()
+        for level in range(recursion_depth):
+            recursive_query = "parent_module_bay__parent_module__" * level
+            query = query | models.Q(**{f"{recursive_query}parent_module_bay__parent_device": self})
+        return qs.filter(query)
+
+    @property
+    def all_console_ports(self):
```
Comment (glennmatthews) [inline nautobot/dcim/models/devices.py:976]: Let's add docstrings to these please. Also might be worth a sentence or two in the Device documentation?

File: nautobot/dcim/tests/test_filters.py [hunk 11]
```diff
@@ -2114,7 +2114,7 @@ class InterfaceTestCase(PathEndpointModelTestMixin, ModularDeviceComponentTestMi
         ("child_interfaces", "child_interfaces__name"),
         ("description",),
         # ("device", "device__id"),  # TODO - InterfaceFilterSet overrides device as a MultiValueCharFilter on name only
-        ("device", "device__name"),
+        # ("device", "device__name"),
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_filters.py:2117]: Comment explaining why this is disabled? Or remove it

File: nautobot/dcim/tests/test_models.py [hunk 12]
```diff
@@ -2872,49 +2904,50 @@ def test_module_infinite_recursion_ancestor(self):
 
     def test_render_component_names(self):
         """Test that creating a Module with components properly renders the {module} and {module.parent} variables."""
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_models.py:2906]: Are the changes to this test relevant to this PR, and are they a net improvement? It feels off to me that we're going from a case where each bay has a distinct position that we can look for in the result to having all of them use the same position `1` that may obscure subtle bugs?

File: nautobot/dcim/models/devices.py (hunk 9 above)
Comment (smk4664) [reply nautobot/dcim/models/devices.py:1806]: I might have chosen the wrong side on a merge conflict. I will change back.

File: nautobot/dcim/forms.py (hunk 3 above)
Comment (smk4664) [reply nautobot/dcim/forms.py:3779]: Correct, I will add it to the TODO.

File: nautobot/dcim/tests/test_models.py (hunk 12 above)
Comment (smk4664) [reply nautobot/dcim/tests/test_models.py:2906]: This might have been introduced when resolving merge conflicts, or something Gary added. I will investigate.

File: nautobot/dcim/models/devices.py (hunk 7 above)
Comment (gsnider2195) [reply nautobot/dcim/models/devices.py:994]: Should be `parent_module__in=self.modules`

File: nautobot/dcim/models/devices.py (hunk 8 above)
Comment (gsnider2195) [reply nautobot/dcim/models/devices.py:1006]: If they both result in the same query, which I suspect they do, I think we should use `Q`.

File: nautobot/dcim/models/devices.py [hunk 13]
```diff
@@ -1394,6 +1444,7 @@ def clean(self):
 #
 
 
+# TODO: Translate comments field from devicetype library, Nautobot doesn't use that field for ModuleType
```
Comment (gsnider2195) [inline nautobot/dcim/models/devices.py:1447]: https://github.com/nautobot/nautobot/issues/5840

File: nautobot/dcim/tests/test_models.py (hunk 12 above)
Comment (smk4664) [reply nautobot/dcim/tests/test_models.py]: I ended up reverting this. I believe this was accidentally added when I tried resolving merge conflicts, and these changes added nothing new, but as pointed out, could obscure potential issues.

File: nautobot/dcim/models/devices.py [hunk 14]
```diff
@@ -955,6 +957,79 @@ def get_children(self):
         """
         return Device.objects.filter(parent_bay__device=self.pk)
 
+    @property
+    def all_modules(self):
```
Comment (glennmatthews) [inline nautobot/dcim/models/devices.py:961]: Would be good to add some unit tests for each of the `all_*` methods?

File: nautobot/dcim/tests/test_models.py [hunk 15]
```diff
@@ -1670,6 +1670,38 @@ def test_primary_ip_validation_logic(self):
...
+            module_type=module_type,
+            parent_module_bay=module_bay,
+            status=status,
+        )
+
+        interface = Interface.objects.create(name="Int1", module=module, status=self.device_status, role=self.intf_role)
+        ips = list(IPAddress.objects.filter(ip_version=4)[:5]) + list(IPAddress.objects.filter(ip_version=6)[:5])
+        interface.add_ip_addresses(ips)
+        device.primary_ip4 = interface.ip_addresses.all().filter(ip_version=4).first()
+        device.primary_ip6 = interface.ip_addresses.all().filter(ip_version=6).first()
```
Comment (glennmatthews) [inline nautobot/dcim/tests/test_models.py:1702]:
```suggestion
        device.primary_ip4 = interface.ip_addresses.all().filter(ip_version=4).first()
        self.assertIsNotNone(device.primary_ip4)
        device.primary_ip6 = interface.ip_addresses.all().filter(ip_version=6).first()
        self.assertIsNotNone(device.primary_ip6)
```

**Review by glennmatthews — APPROVED** [review]

**Comment by glennmatthews** [comment]
This was my fix to the CI failure you're showing:

https://github.com/nautobot/nautobot/pull/5845/files#diff-bf29c86a684ab0f16aa0ed3546457cba0783d5e80909f51710ea1fe1fc6818f2L221-R221
